File: solutions/006-Longest-Substring-With-Without-Repetition/solution.cs
using System;
using System.Collections.Generic;

public class Solution
{
    // Function to find the longest substring without repeating characters
    public int LongestSubstringWithoutRepetition(string s)
    {
        int n = s.Length;
        int maxLength = 0;
        Dictionary<char, int> charIndexMap = new Dictionary<char, int>();
        int start = 0;

        for (int end = 0; end < n; end++)
        {
            char currentChar = s[end];

            if (charIndexMap.ContainsKey(currentChar))
            {
                start = Math.Max(start, charIndexMap[currentChar] + 1);
            }

            charIndexMap[currentChar] = end;
            maxLength = Math.Max(maxLength, end - start + 1);
        }

        return maxLength;
    }

    // Function to find the longest substring with at most two repeating characters
    public int LongestSubstringWithTwoRepeatingCharacters(string s)
    {
        int n = s.Length;
        int maxLength = 0;
        Dictionary<char, int> charCountMap = new Dictionary<char, int>();
        int start = 0;
        int distinctCount = 0;

        for (int end = 0; end < n; end++)
        {
            char currentChar = s[end];

            if (charCountMap.ContainsKey(currentChar))
            {
                charCountMap[currentChar]++;
            }
            else
            {
                charCountMap[currentChar] = 1;
                distinctCount++;
            }

            while (distinctCount > 2)
            {
                char startChar = s[start];
                charCountMap[startChar]--;
                if (charCountMap[startChar] == 0)
                {
                    charCountMap.Remove(startChar);
                    distinctCount--;
                }
                start++;
            }

            maxLength = Math.Max(maxLength, end - start + 1);
        }

        return maxLength;
    }
}

// Example usage
public class Program
{
    public static void Main(string[] args)
    {
        Solution solution = new Solution();
        string s = "abcabcbb";
        int resultWithoutRepetition = solution.LongestSubstringWithoutRepetition(s);
        int resultWithTwoRepeating = solution.LongestSubstringWithTwoRepeatingCharacters(s);
        Console.WriteLine(resultWithoutRepetition); // Expected output: 3
        Console.WriteLine(resultWithTwoRepeating); // Expected output: 7
    }
}

File: solutions/006-Longest-Substring-With-Without-Repetition/solution.py
from typing import Dict

class Solution:
    def longest_substring_without_repetition(self, s: str) -> int:
        n = len(s)
        max_length = 0
        char_index_map: Dict[str, int] = {}
        start = 0

        for end in range(n):
            current_char = s[end]

            if current_char in char_index_map:
                start = max(start, char_index_map[current_char] + 1)

            char_index_map[current_char] = end
            max_length = max(max_length, end - start + 1)

        return max_length

    def longest_substring_with_two_repeating_characters(self, s: str) -> int:
        n = len(s)
        max_length = 0
        char_count_map: Dict[str, int] = {}
        start = 0
        distinct_count = 0

        for end in range(n):
            current_char = s[end]

            if current_char in char_count_map:
                char_count_map[current_char] += 1
            else:
                char_count_map[current_char] = 1
                distinct_count += 1

            while distinct_count > 2:
                start_char = s[start]
                char_count_map[start_char] -= 1
                if char_count_map[start_char] == 0:
                    del char_count_map[start_char]
                    distinct_count -= 1
                start += 1

            max_length = max(max_length, end - start + 1)

        return max_length

# Example usage
if __name__ == "__main__":
    solution = Solution()
    s = "abcabcbb"
    result_without_repetition = solution.longest_substring_without_repetition(s)
    result_with_two_repeating = solution.longest_substring_with_two_repeating_characters(s)
    print(result_without_repetition)  # Expected output: 3
    print(result_with_two_repeating)  # Expected output: 7
